internal/workload/podspec_updates.go

```diff
@@ -613,6 +640,75 @@ func (s *updateState) addHealthCheck(_ *cloudsqlapi.AuthProxyWorkload, c *corev1
 	return cliArgs
 }
 
+func (s *updateState) addVolumeMount(p *cloudsqlapi.AuthProxyWorkload, is *cloudsqlapi.InstanceSpec, m corev1.VolumeMount, v corev1.Volume) {
+	key := dbInst(p.Namespace, p.Name, is.ConnectionString)
+	vol := &managedVolume{
+		Instance:    key,
+		Volume:      v,
+		VolumeMount: m,
+	}
+
+	for i, mount := range s.mods.VolumeMounts {
+		if mount.Instance == key {
+			s.mods.VolumeMounts[i] = vol
+			return
+		}
+		if mount.VolumeMount.MountPath == vol.VolumeMount.MountPath {
+			// avoid adding volume mounts with redundant MountPaths,
+			// just the first one is enough.
+			return
+		}
+	}
+	s.mods.VolumeMounts = append(s.mods.VolumeMounts, vol)
+}
+
+// applyContainerVolumes applies all the VolumeMounts to this container.
+func (s *updateState) applyContainerVolumes(c *corev1.Container) {
+	nameAccessor := func(v corev1.VolumeMount) string {
+		return v.Name
+	}
+	thingAccessor := func(v *managedVolume) corev1.VolumeMount {
+		return v.VolumeMount
+	}
+	c.VolumeMounts = applyVolumeThings[corev1.VolumeMount](s, c.VolumeMounts, nameAccessor, thingAccessor)
+}
+
+// applyVolumes applies all volumes to this PodSpec.
+func (s *updateState) applyVolumes(ps *corev1.PodSpec) {
+	nameAccessor := func(v corev1.Volume) string {
+		return v.Name
+	}
+	thingAccessor := func(v *managedVolume) corev1.Volume {
+		return v.Volume
+	}
+	ps.Volumes = applyVolumeThings[corev1.Volume](s, ps.Volumes, nameAccessor, thingAccessor)
+}
+
+// applyVolumeThings implements complex reconcile logic that is duplicated for both
```

Comment:
Could we elaborate on what "complex" means here?

Reply:
I rewrote the comment to be more precise.